editor/xml_feature.cpp

```diff
@@ -0,0 +1,214 @@
+char const * const XMLFeature::kHouseNumber = "addr:housenumber";
+char const * const XMLFeature::kDefaultName = "name";
+char const * const XMLFeature::kLocalName = "name:";
+char const * const XMLFeature::kDefaultLang =
+    StringUtf8Multilang::GetLangByCode(StringUtf8Multilang::DEFAULT_CODE);;
+
+XMLFeature::XMLFeature()
+{
+  m_document.append_child("node");
+}
+
+XMLFeature::XMLFeature(string const & xml)
+{
+  m_document.load(xml.data());
+  ValidateNode(GetRootNode());
+}
+
+XMLFeature::XMLFeature(pugi::xml_document const & xml)
+{
+  m_document.reset(xml);
+  ValidateNode(GetRootNode());
+}
+
+XMLFeature::XMLFeature(pugi::xml_node const & xml)
+{
+  m_document.reset();
+  m_document.append_copy(xml);
+  ValidateNode(GetRootNode());
+}
+
+void XMLFeature::Save(ostream & ost) const
+{
+  m_document.save(ost, "  ", pugi::format_indent_attributes);
+}
+
+m2::PointD XMLFeature::GetCenter() const
+{
+  return PointFromLatLon(GetRootNode());
+}
+
+void XMLFeature::SetCenter(m2::PointD const & mercatorCenter)
+{
+  SetAttribute("lat", strings::to_string_dac(MercatorBounds::YToLat(mercatorCenter.y),
+                                             kLatLonTolerance));
+  SetAttribute("lon", strings::to_string_dac(MercatorBounds::XToLon(mercatorCenter.x),
+                                             kLatLonTolerance));
+}
+
+string XMLFeature::GetName(string const & lang) const
+{
+  auto const suffix = (lang == kDefaultLang || lang.empty()) ? "" : ":" + lang;
+  return GetTagValue("name" + suffix);
+}
+
+string XMLFeature::GetName(uint8_t const langCode) const
+{
+  return GetName(StringUtf8Multilang::GetLangByCode(langCode));
+}
+
+void XMLFeature::SetName(string const & name)
+{
+  SetName(kDefaultLang, name);
+}
+
+void XMLFeature::SetName(string const & lang, string const & name)
+{
+  auto const suffix = (lang == kDefaultLang || lang.empty()) ? "" : ":" + lang;
+  SetTagValue("name" + suffix, name);
+}
+
+void XMLFeature::SetName(uint8_t const langCode, string const & name)
+{
+  SetName(StringUtf8Multilang::GetLangByCode(langCode), name);
+}
+
+string XMLFeature::GetHouse() const
+{
+  return GetTagValue(kHouseNumber);
+}
+
+void XMLFeature::SetHouse(string const & house)
+{
+  SetTagValue(kHouseNumber, house);
+}
+
+time_t XMLFeature::GetModificationTime() const
+{
+  auto const node = GetRootNode();
+  return my::StringToTimestamp(node.attribute("timestamp").value());
+}
+
+void XMLFeature::SetModificationTime(time_t const time)
+{
+  SetAttribute(kLastModified, my::TimestampToString(time));
+}
+
+bool XMLFeature::HasTag(string const & key) const
+{
+  return FindTag(m_document, key);
+}
+
+bool XMLFeature::HasAttribute(string const & key) const
+{
+  return GetRootNode().attribute(key.data());
+}
+
+bool XMLFeature::HasKey(string const & key) const
+{
+  return HasTag(key) || HasAttribute(key);
+}
+
+string XMLFeature::GetTagValue(string const & key) const
+{
+  auto const tag = FindTag(m_document, key);
```

Comment:
FindTag performs more complicated logic than GetRootNode.
So I would like to leave it as is.